### PR TITLE
Reports Feature Refactor + UI

### PR DESCRIPTION
**Pull request by cconard96**
| Q             | A
| ------------- | ---
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | -

This wasn't meant to be a rewrite of the report feature, but the required refactoring is extensive.
1. Removal of data retrieval and UI from front files.
2. Separation of data retrieval and UI (Prerequisite for HL API).
3. Twig templates for UI.
4. Addition of some generic report rendering methods for the various types of reports.
5. Cleanup of the display of some reports.
   - Default report had counts with potentially confusing labels. It would list all assets by type, OS by name and then asset types (ComputerType, etc) by the itemtype name (Computer) in the same list. Now, the labels are like "Assets > Computers", "Operating Systems > Fedora Linux 38", "Types > Computers > QEMU", etc.
6. Conversion of forms to use `GET` instead of `POST` (both were used in different reports).
7. Unification of the report and its criteria (some reports have a URL just for the criteria and another just for the result).

### CODE REVIEW

**Comment by cconard96** [comment]
Some help is needed regarding the graphs that are supposed to show for the two Infocom reports.

File: templates/components/datatable.html.twig [hunk 1]
```diff
@@ -76,9 +76,9 @@
                     {% set super_header_label = super_header is array ? super_header['label'] : super_header %}
                     {% set super_header_raw = super_header is array ? super_header['is_raw'] : false %}
                     <tr>
-                        <th colspan="{{ total_cols }}">
+                        {% if super_header_raw != 'th_elements' %}<th colspan="{{ total_cols }}">{% endif %}
```
Comment (trasher) [inline templates/components/datatable.html.twig:79]: It sound strange to provide th elements as raw html here.

Comment (cconard96) [reply templates/components/datatable.html.twig:79]: The alternative is refactoring the datatable template to allow a lot more customization for edge-cases.
I suppose it needs done eventually if we want to get rid of the HTMLTable* classes.

**Comment by trasher** [comment]
I think I've never used that part before; I really do not know how it should work.

On default report, I have quite a different result:
on current main:
image

On your PR:
image

Event if the new display may be correct (as I do not understand the old one at all); at least the last line seems strange.

Also, please rebase, I had to switch between main and your branch, and I had to rerun dependencies install each time.

**Comment by cconard96** [comment]
> Event if the new display may be correct (as I do not understand the old one at all); at least the last line seems strange.
> 
> Also, please rebase, I had to switch between main and your branch, and I had to rerun dependencies install each time.

The new format still isn't great, but better. The last line indicates that it is the count for Computers with no type (----- being the display value for an empty selection). All the lines without a count that were in the old report are skipped in the new one to reduce noise.

**Comment by trasher** [comment]
"----" is the empty default for dropdowns, I'm really not sure that make sense for anyone outside this context -but I understand it was present before.
Maybe it's the good time to change for something more explicit?

**Comment by cconard96** [comment]
> "----" is the empty default for dropdowns, I'm really not sure that make sense for anyone outside this context -but I understand it was present before. Maybe it's the good time to change for something more explicit?

N/A is used in other places. Perhaps that but also in italics.

**Comment by trasher** [comment]
> > "----" is the empty default for dropdowns, I'm really not sure that make sense for anyone outside this context -but I understand it was present before. Maybe it's the good time to change for something more explicit?
> 
> N/A is used in other places. Perhaps that but also in italics?

It seems more easy to understand to me, yes

**Comment by trasher** [comment]
On `report.infocom.php`, I have the following error: 
```
[2024-05-16 12:58:34] glpiphplog.CRITICAL:   *** Uncaught Exception TypeError: Unsupported operand types: int + string in /var/www/html/private/glpi/src/Report.php at line 1652
```

**Comment by cconard96** [comment]
> > On `report.infocom.php`, I have the following error:
> > ```
> > [2024-05-16 12:58:34] glpiphplog.CRITICAL:   *** Uncaught Exception TypeError: Unsupported operand types: int + string in /var/www/html/private/glpi/src/Report.php at line 1652
> > ```
> 
> Problem is `$item['anv']` can be `'-'`

Ah. The phpdoc for `Infocom::Amort` is wrong.

**Comment by trasher** [comment]
> Ah. The phpdoc for `Infocom::Amort` is wrong.

Well.. Most of "amort" part is wrong :D

What I do not understand is I have not that issue on main, but I also do not have the "ANV" column displayed:
image

**Comment by cconard96** [comment]
> What I do not understand is I have not that issue on main, but I also do not have the "ANV" column displayed

In French the label is VNC.

**Comment by trasher** [comment]
> In French the label is VNC.

Indeed, I did not pay attention lang has switched (I usually work in english :D).

Another issue: I do not have the search form (Start/End date) with "Other financial and ..." (no error in logs).

**Comment by trasher** [comment]
End of my tests:
- Loan report: seems I do not have any data, and I do no know what's it's related to,
- Status report: OK!

**Comment by trasher** [comment]
> Loan report is related to reservations (bad name).

OK, I have an SQL error (empty in are not allowed) trying to add a new reservation :/

**Comment by trasher** [comment]
> > Loan report is related to reservations (bad name).
> 
> OK, I have an SQL error (empty in are not allowed) trying to add a new reservation :/ I'll try that later

ON another DB, I do not have the problem. Loan report is OK too.

**Comment by trasher** [comment]
So, remaining issues:

- [ ] I do not have the search form (Start/End date) with "Other financial and ..." (no error in logs).
- [ ] Network report, "By hardware" I have data in the table for all equipements I can choose, but on your PR, I always have "No data".
- [ ] Network report, "By hardware" searched entry is not selected

**Comment by cconard96** [comment]
> So, remaining issues:
> 
>     * [ ]  I do not have the search form (Start/End date) with "Other financial and ..." (no error in logs).
Report not implemented at all because it had no tabular data. The only thing it showed are the graphs which already seemed broken and I have no idea how they are supposed to work.

>     * [ ]  Network report, "By hardware" I have data in the table for all equipements I can choose, but on your PR, I always have "No data".
>     * [ ]  Network report, "By hardware" searched entry is not selected

Maybe fixed.

>     * [ ]  when changing report from the dropdown, we should not rely on configured URL

Fixed

**Comment by trasher** [comment]
> > So, remaining issues:
> > ```
> > * [ ]  I do not have the search form (Start/End date) with "Other financial and ..." (no error in logs).
> > ```
> 
> Report not implemented at all because it had no tabular data. The only thing it showed are the graphs which already seemed broken and I have no idea how they are supposed to work.

Unfortunately, I cannot help :/

All other issues are fixed; LGTM.

**Comment by trasher** [comment]
If the "Other financial and ..." report is broken; I propose to just drop it. ping @orthagh?

**Comment by orthagh** [comment]
If the thing is broken since a long time, you may remove it.
I would want a bit more investigation on the history of the report

**Comment by trasher** [comment]
OK, it seems my first tests were wrong. I've just tested again on a 10.0/bf fresh install; and graphs are working

image

**Comment by trasher** [comment]
I still have an empty page for "Other financial" report

**Comment by trasher** [comment]
I cannot select any time range ^^

**Comment by cconard96** [comment]
I guess there should be a PHP error in the logs for this.

**Comment by trasher** [comment]
Unfortunately no; there is absolutely nothing in logs :( I'll try again today; I'll take care of remonving all possible cache to be sure.

**Comment by trasher** [comment]
OK, it's working as expected on another instance; no idea what was wrong with the one I initially used :)

**Review by trasher — APPROVED** [review]

**Pull request merged**